### ISSUE TITLE
Socketcluster client and nativescript not playing nice

### ISSUE THREAD

**Issue by jcjmcclean** [SocketCluster/socketcluster#496]
I've got a nativescript project which needs to interface with a socketcluster server. I've already created a proof of concept in angular (without nativescript), which works perfectly using socketcluster browser client - but I can't get the nativescript angular android app to work with socketcluster (using the same browser client). 

Nativescript doesn't support websockets out of the box, there's a plugin [nativescript-websockets](https://github.com/NathanaelA/nativescript-websockets) which enables websocket support. I created another proof of concept in nativescript using a vanilla websocket server and the nativescript-websockets plugin which worked perfectly again (connected, received and sent messages + disconnected). 

So now I know that socketcluster.js browser client works with angular and nativescript-websockets works with nativescript. I need to combine the 2 to enable my nativescript app to connect to the socketcluster server...

I edited the socketcluster.js browser client to switch out line `1304` (`WebSocket = _dereq_('ws');`) for the nativescript websockets library `WebSocket = require('nativescript-websockets');` hoping that would enable the socketcluster library to work with nativescript.

The response I'm getting from that is error code `4007: 'ERROR SocketProtocolError: Client connection establishment timed out'`.

I can connect to the server with the same settings from the angular app and vanilla websockets works on nativescript, so something isn't playing nice between socketcluster client and nativescript-websockets. 

**WIRESHARK**
I setup wireshark to listen for websocket packets, so I can see if/when the socketcluster connection is created. It works great with the angular app and with vanilla websockets on nativescript but sees nothing at all happening when I'm using socketcluster client in the nativescript app, it's like the websocket connection isn't even being attempted.

**TL;DR**
Trying to get socketcluster client working on a nativescript application using nativescript-websockets. Getting timeout errors. Same code works in angular app.

**Question:**
Is there any debugging available within socketcluster client?

**Comment by jcjmcclean** [comment]
For reference links to other issues also opened:
https://github.com/SocketCluster/socketcluster-client/issues/127
https://github.com/NathanaelA/nativescript-websockets/issues/67

**Comment by MegaGM** [comment]
@jcjmcclean wrote:
> I edited the socketcluster.js browser client to switch out line 1304 (WebSocket = _dereq_('ws');) for the nativescript websockets library WebSocket = require('nativescript-websockets'); hoping that would enable the socketcluster library to work with nativescript.

> The response I'm getting from that is error code 4007: 'ERROR SocketProtocolError: Client connection establishment timed out'.

---


### TL;DR
SocketCluster works fine with NativeScript.  
You have required `NativeWebSockets` instead of `BrowserWebSockets` from `nativescript-websockets` plugin.

--- 

Why do I think so? Because I've created a NativeScript (v6.1.2) app, have installed [SocketCluster JS client](https://github.com/SocketCluster/socketcluster-client/blob/master/socketcluster.js) `npm i -S socketcluster-client`, and have copy-pasted [some code from SC sample application](https://github.com/SocketCluster/socketcluster/blob/master/sample/public/index.html#L72-L117) into the app.

And I got an error:
```sh
TypeError: WebSocket is not a constructor
```
Which is fine, because, as you have mentioned before, NativeScript has no WebSockets support out of the box.

So I've installed `nativescript-websockets` plugin.
And everything just works as it intended to.

Then, out of curiosity, I tried to do what you did, to change the same lines of code in `socketcluster-client`
```js
// from
WebSocket = _dereq_('ws');
// to
WebSocket = require('nativescript-websockets');
```
And everything still works fine. Because we should never actually hit the second branch in [the `if` statement](https://github.com/SocketCluster/socketcluster-client/blob/master/lib/sctransport.js#L7-L17), within which you changed the code. That branch is for launching in environments like Node.js But with NativeScript we need to hit the first branch, which is for web browsers.

During startup, SocketCluster JS client [looks for `global.WebSocket`](https://github.com/SocketCluster/socketcluster-client/blob/master/lib/sctransport.js#L7) constructor. Since you hit the second branch in the `if` statement, instead of the first, then, apparently, you have missed requiring it.

So I've commented the line where I require `nativescript-websockets` into my NativeScript app and, quite predictably, got the same error as you got:

```sh
4007: 'ERROR SocketProtocolError: Client connection establishment timed out'.
```

---


Essentially you didn't have to modify SocketCluster source code at all, you just have missed the second step in a checklist:

1. Install `nativescript-websockets` plugin
```sh
$ tns plugin add nativescript-websockets
```

2. Require `nativescript-websockets` plugin, so `global.WebSockets` would have been in place for SocketCluster to pick up

3. Require SocketCluster JS client

4. Then you're free to use SocketCluster as if you would have been in a web browser

```js
// JS
require('nativescript-websockets');
var socketCluster = require('socketcluster-client/socketcluster.min');
var socket = socketCluster.create(options);
```
```ts
// TS
import 'nativescript-websockets'
import socketCluster from 'socketcluster-client/socketcluster.min'
const socket = socketCluster.create(options)
```
